src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -404,9 +404,38 @@ internal override bool ApplyNullableTransforms(byte defaultTransformFlag, Immuta
             return true;
         }
 
-        internal override TypeSymbol SetNullabilityForReferenceTypes(Func<TypeWithAnnotations, TypeWithAnnotations> transform)
+        internal override TypeSymbol SetObliviousNullabilityForReferenceTypes()
         {
-            return WithElementType(transform(ElementTypeWithAnnotations));
+            // The language supports deeply nested arrays, up to 10,000+ instances. This means we can't implemented a head
+            // recursive solution here. Need to take an iterative approach to building up the annotations here.
```

Comment:
>Need to take an iterative approach to building up the annotations here. [](start = 40, length = 71)

Why are we concern with a recursion only within arrays, what about a recursion for pointer types or generic types? Also, it looks like we are fine with relying on recursion for ```AddNullableTransforms``` and ```ApplyNullableTransforms```. Why ```SetObliviousNullabilityForReferenceTypes``` is more important? Are we addressing a real problem?

Reply:
> Why SetObliviousNullabilityForReferenceTypes is more important?

The bug filed was for this case. If we want to expand out to the other cases that's fine to. 

> what about a recursion for pointer types or generic types?

It's the level of nesting that is the concern. The compiler supports nesting levels of ~500 for generics. That's a reasonable depth for recursive solutions. The tests for this array case indicate we support arrays at a depth of 10,000+ which is not suitable for recursion (blows the stack every time). There are no tests I'm aware of for pointers that suggest we support that level of nesting.

Reply:
> The bug filed was for this case. If we want to expand out to the other cases that's fine to. 

I think than in this case we should be concerned about any API that takes recursive approach over arrays. If we don't have a targeted test, that doesn't mean we don't have a problem. I personally not convinced that we have to handle 10000+ deep arrays.

---
In reply to: [274079994](https://github.com/dotnet/roslyn/pull/34891#discussion_r274079994) [](ancestors = 274079994)

Reply:
Here is the original customer bug for the test case. They do indeed have arrays of incredible depth in their code base  (see Class1pf3fbcd0.nxt.cs in the attachment for an example). It's to validate their serialization framework (from the look of it). 

http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/531552

Reply:
> Here is the original customer bug for the test case. They do indeed have arrays of incredible depth in their code base  (see Class1pf3fbcd0.nxt.cs in the attachment for an example). It's to validate their serialization framework (from the look of it). 

There are always people that push things to the extreme. If we think we absolutely have to support them, then I think we should be concerned with ```AddNullableTransforms```, ```ApplyNullableTransforms``` and ```MergeNullability``` as well.

---
In reply to: [274092328](https://github.com/dotnet/roslyn/pull/34891#discussion_r274092328) [](ancestors = 274092328)